converters/google/convert.go

```diff
@@ -302,7 +302,7 @@ func (c *Converter) addCreateOrUpdate(rc *tfjson.ResourceChange) error {
 				if mapper.mergeCreateUpdate == nil {
 					// If a merge function does not exist ignore the asset and return
 					// a checkable error.
-					return fmt.Errorf("asset type %s: asset name %s %w", converted.Type, converted.Name, ErrDuplicateAsset)
```

Comment:
Shouldn't the `%w` directive have wrapped this? As described at https://go.googlesource.com/proposal/+/master/design/29934-error-values.md#changes-to.

Instead of `errors.Cause` from `github.com/pkg/errors` can we use [`errors.Is` from the system library](https://golang.org/pkg/errors/#example_Is) or is it significantly semantically different?

Reply:
@rileykarson very good question.

After some testing and GitHub exploration we got:

| _Creation_  \\ _Checking_ |  **errors.Is**(github.com/golang/go) | **errors.Cause**(github.com/pkg) |
|---|:-:|:-:|
| **fmt.Errorf**(github.com/golang/go) | OK | **FAIL** |
| **errors.Wrapf**(github.com/pkg) | OK \* | OK |


The  `%w ` directive does wrap the error as indicted in the link, the issue is that:


#### The `fmt.Errorf` + `errors.Is` pair uses a chain of `Unwrap()` calls to navigate the error chain

https://github.com/golang/go/blob/go1.14.15/src/fmt/errors.go#L17

```go
func Errorf(format string, a ...interface{}) error {
	p := newPrinter()
	p.wrapErrs = true
	p.doPrintf(format, a)
	s := string(p.buf)
	var err error
	if p.wrappedErr == nil {
		err = errors.New(s)
	} else {
		err = &wrapError{s, p.wrappedErr}
	}
	p.free()
	return err
}
```


https://github.com/golang/go/blob/go1.14.15/src/errors/wrap.go#L14

```go
func Unwrap(err error) error {
	u, ok := err.(interface {
		Unwrap() error
	})
	if !ok {
		return nil
	}
	return u.Unwrap()
}
```


#### The `errors.Wrapf` + `errors.Cause` pair uses a chain of `Cause()` calls to navigate the error chain

https://github.com/pkg/errors/blob/v0.9.1/errors.go#L201

```go
func Wrapf(err error, format string, args ...interface{}) error {
	if err == nil {
		return nil
	}
	err = &withMessage{
		cause: err,
		msg:   fmt.Sprintf(format, args...),
	}
	return &withStack{
		err,
		callers(),
	}
}
```


https://github.com/pkg/errors/blob/v0.9.1/errors.go#L275
```go
func Cause(err error) error {
	type causer interface {
		Cause() error
	}

	for err != nil {
		cause, ok := err.(causer)
		if !ok {
			break
		}
		err = cause.Cause()
	}
	return err
}
```

#### **\*** The `errors.Wrapf`  + `errors.Is` pair  works because struct `withMessage` implements `Unwrap()` 

https://github.com/pkg/errors/blob/v0.9.1/errors.go#L248

```go
func (w *withMessage) Unwrap() error { return w.cause }
```

Considering this context and that `github.com/golang/go/errors` is not used anywhere in terraform-validator, but we still have three cases of usage of `fmt.Errorf` in the code:

https://github.com/GoogleCloudPlatform/terraform-validator/blob/6cc8af9992499c25aeea698d71d944140712a82c/converters/google/convert.go#L196

https://github.com/GoogleCloudPlatform/terraform-validator/blob/6cc8af9992499c25aeea698d71d944140712a82c/converters/google/convert.go#L206

https://github.com/GoogleCloudPlatform/terraform-validator/blob/6cc8af9992499c25aeea698d71d944140712a82c/converters/google/convert.go#L345

which were introduced as  part of PR https://github.com/GoogleCloudPlatform/terraform-validator/pull/180 (which cuts and restarts the error chain at this point)

I think that:

1) This commit https://github.com/GoogleCloudPlatform/terraform-validator/commit/41dd2648e3c544793bd3501108aa564f83fcb65a cloud be reverted in this Pull Request
2) A new issue could be open to plan how to migrate:

- `errors.Cause()` => `errors.Is()` 
- `errors.Wrapf()` => `fmt.Errorf()`

In the whole terraform-validator at the same time.



@rileykarson @melinath what do you think?

Reply:
Thanks for the thorough investigation! It looks like ErrDuplicateAsset is the only case where one of the changes from https://github.com/GoogleCloudPlatform/terraform-validator/commit/41dd2648e3c544793bd3501108aa564f83fcb65a could result in this kind of bug; the other cases are not directly being checked anywhere. Based on the discussion, my preference would be to keep `fmt.Errorf()`and switch the check for ErrNoConversion to use `errors.Is()`. We can then have the conversation about the remaining usages of `errors.Wrapf()` / `errors.Cause()` separately.

Reply:
PR updated to use `errors.Is` + `fmt.Errorf`